uarray/_backend.py

```diff
@@ -62,13 +79,76 @@ def pickle_function(func):
             "Can't pickle {}: it's not the same object as {}".format(func, test)
         )
 
-    return unpickle_function, (mod_name, qname)
+    return unpickle_function, (mod_name, qname, self_)
+
+
+def pickle_state(state):
+    return _uarray._BackendState._unpickle, state._pickle()
+
+
+def pickle_set_backend_context(ctx):
+    return _SetBackendContext, ctx._pickle()
+
+
+def pickle_skip_backend_context(ctx):
+    return _SkipBackendContext, ctx._pickle()
 
 
 copyreg.pickle(_Function, pickle_function)
+copyreg.pickle(_uarray._BackendState, pickle_state)
+copyreg.pickle(_SetBackendContext, pickle_set_backend_context)
+copyreg.pickle(_SkipBackendContext, pickle_skip_backend_context)
 atexit.register(_uarray.clear_all_globals)
 
 
+def get_state():
+    """
+    Returns an opaque object containing the current state of all the backends.
+
+    Can be used for synchronization between threads/processes.
+
+    See Also
+    --------
+    set_state
+        Sets the state returned by this function.
+    """
+    return _uarray.get_state()
+
+
+@contextlib.contextmanager
+def reset_state():
```

Comment:
I think that if you're currently using the `global_domain_map` then `reset_state` should copy it into your `thread_global_domain_map` and use that within the current context. Basically, it should be equivalent to:
```python
with ua.set_state(ua.get_state()):
    # ...
```
Which I admit looks a bit weird but otherwise we aren't resetting the global context. (I think this is why tests are failing by the way.